### PR TITLE
fix: duplicate resize images logic of qwen-vl series models

### LINKED ISSUE #12390: [Bug] Qwen-VL resize outside image-processor will lead gap for VIT and total model's output
### Checklist

- [ ] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

In sglang/python/sglang/srt/multimodal/processors/qwen_vl.py's line 285
resize image will be executed.  While in processor.__call__, resize will be executed again.
Two problems will be leaded in this case:
1. PIL's resize is time-costing operation, which will bring an extra time-cost. 
2. currently, resize outside use a fixed parameters, 

IMAGE_FACTOR = 28
MIN_PIXELS = 4 * 28 * 28
for Qwen2.5-VL's case, it's ok. while for Qwen3-VL , since the patch size is 16 , which is dirrent with 14 here. This fixed parameters will make an image fisrtly resize as 14, then resize as 16. As a result, the VIT's input and output will have some mismatch. In some case， the visiontoken's num will change.

With resize outside:
img-precessor output :
pixel_values.shape torch.Size([640, 1536])
pixel_values mean -0.291015625
pixel_values max 1.0

VIT output
image_embeds.shape torch.Size([160, 16384])
image_embeds mean -1.800060272216797e-05
image_embeds max 4.5

LLM output:
the last line of logits tensor([12.5000, 17.6250, 15.5000,  ..., -4.3750, -4.3750, -4.3750],
       device='cuda:0')
logits mean -2.0931811332702637
logits max 38.0

Without resize outside:

image-precessor output:
pixel_values.shape torch.Size([704, 1536])
pixel_values mean -0.291015625
pixel_values max 1.0

VIT output:
image_embeds.shape torch.Size([176, 16384])
image_embeds mean 2.9921531677246094e-05
image_embeds max 4.375

LLM output:
the last line of logits tensor([12.4375, 17.7500, 15.1875,  ..., -4.4688, -4.4688, -4.4688],
       device='cuda:0')
logits mean -2.2209458351135254
logits max 38.0

even the logits max keep the samec here ,maybe some case the result will be different.

### Reproduction

Qwen2.5-VL

### Environment

/usr/local/lib/python3.12/dist-packages/torch/cuda/__init__.py:63: FutureWarning: The pynvml package is deprecated. Please install nvidia-ml-py instead. If you did not install pynvml directly, please report this to the maintainers of the package that installed pynvml for you.
  import pynvml  # type: ignore[import]
Python: 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA H20
GPU 0,1,2,3,4,5,6,7 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.9, V12.9.86
CUDA Driver Version: 535.161.08
PyTorch: 2.8.0+cu129
sglang: 0.5.4.post1
sgl_kernel: 0.3.16.post4
flashinfer_python: 0.4.1
triton: 3.4.0
transformers: 4.57.1
torchao: 0.9.0
numpy: 1.26.4
aiohttp: 3.13.0
fastapi: 0.118.3
hf_transfer: 0.1.9
huggingface_hub: 0.35.3
interegular: 0.3.3
modelscope: 1.30.0
orjson: 3.11.3
outlines: 0.1.11
packaging: 25.0
psutil: 7.1.0
pydantic: 2.12.0
python-multipart: 0.0.20
pyzmq: 27.1.0
uvicorn: 0.37.0
uvloop: 0.21.0
vllm: Module Not Found
xgrammar: 0.1.25
openai: 2.6.1
tiktoken: 0.12.0
anthropic: 0.69.0
litellm: Module Not Found
decord2: 2.0.0
NVIDIA Topology:
	GPU0	GPU1	GPU2	GPU3	GPU4	GPU5	GPU6	GPU7	NIC0	NIC1	NIC2	NIC3	NIC4	NIC5	NIC6	NIC7	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	NV18	NV18	NV18	NV18	NV18	NV18	NV18	PIX	NODE	NODE	NODE	SYS	SYS	SYS	SYS	0-47,96-143	0		N/A
GPU1	NV18	 X 	NV18	NV18	NV18	NV18	NV18	NV18	NODE	PIX	NODE	NODE	SYS	SYS	SYS	SYS	0-47,96-143	0		N/A
GPU2	NV18	NV18	 X 	NV18	NV18	NV18	NV18	NV18	NODE	NODE	PIX	NODE	SYS	SYS	SYS	SYS	0-47,96-143	0		N/A
GPU3	NV18	NV18	NV18	 X 	NV18	NV18	NV18	NV18	NODE	NODE	NODE	PIX	SYS	SYS	SYS	SYS	0-47,96-143	0		N/A
GPU4	NV18	NV18	NV18	NV18	 X 	NV18	NV18	NV18	SYS	SYS	SYS	SYS	PIX	NODE	NODE	NODE	48-95,144-191	1		N/A
GPU5	NV18	NV18	NV18	NV18	NV18	 X 	NV18	NV18	SYS	SYS	SYS	SYS	NODE	PIX	NODE	NODE	48-95,144-191	1		N/A
GPU6	NV18	NV18	NV18	NV18	NV18	NV18	 X 	NV18	SYS	SYS	SYS	SYS	NODE	NODE	PIX	NODE	48-95,144-191	1		N/A
GPU7	NV18	NV18	NV18	NV18	NV18	NV18	NV18	 X 	SYS	SYS	SYS	SYS	NODE	NODE	NODE	PIX	48-95,144-191	1		N/A
NIC0	PIX	NODE	NODE	NODE	SYS	SYS	SYS	SYS	 X 	NODE	NODE	NODE	SYS	SYS	SYS	SYS
NIC1	NODE	PIX	NODE	NODE	SYS	SYS	SYS	SYS	NODE	 X 	NODE	NODE	SYS	SYS	SYS	SYS
NIC2	NODE	NODE	PIX	NODE	SYS	SYS	SYS	SYS	NODE	NODE	 X 	NODE	SYS	SYS	SYS	SYS
NIC3	NODE	NODE	NODE	PIX	SYS	SYS	SYS	SYS	NODE	NODE	NODE	 X 	SYS	SYS	SYS	SYS
NIC4	SYS	SYS	SYS	SYS	PIX	NODE	NODE	NODE	SYS	SYS	SYS	SYS	 X 	NODE	NODE	NODE
NIC5	SYS	SYS	SYS	SYS	NODE	PIX	NODE	NODE	SYS	SYS	SYS	SYS	NODE	 X 	NODE	NODE
NIC6	SYS	SYS	SYS	SYS	NODE	NODE	PIX	NODE	SYS	SYS	SYS	SYS	NODE	NODE	 X 	NODE
NIC7	SYS	SYS	SYS	SYS	NODE	NODE	NODE	PIX	SYS	SYS	SYS	SYS	NODE	NODE	NODE	 X

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_1
  NIC1: mlx5_2
  NIC2: mlx5_3
  NIC3: mlx5_4
  NIC4: mlx5_5
  NIC5: mlx5_6
  NIC6: mlx5_7
  NIC7: mlx5_8

ulimit soft: 1048576

### PR DESCRIPTION
**Pull request by yangsijia-celina**
## Motivation

fix https://github.com/sgl-project/sglang/issues/12390 and https://github.com/sgl-project/sglang/issues/11896

## Modifications

1. set `MIN_PIXELS`, `MAX_PIXELS`, `IMAGE_FACTOR` according to model's config
2. remove `resize_image` logic in qwen-vl series processors because resize will always be done in transformers, also remove other useless functions and imports.

## Accuracy Tests

The results of test_vlm_models.py
before fix:
image

after fix:
image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/multimodal/processors/qwen_vl.py [hunk 1]
```diff
@@ -250,10 +251,20 @@ def __init__(self, hf_config, server_args, _processor, *args, **kwargs):
...
+        if hasattr(processor, "image_processor") and isinstance(
+            processor.image_processor, BaseImageProcessorFast
+        ):
+            self.MIN_PIXELS = processor.image_processor.size["shortest_edge"]
+            self.MAX_PIXELS = processor.image_processor.size["longest_edge"]
         self.MAX_RATIO = 200
+
+        self.PATCH_SIZE = hf_config.vision_config.patch_size
+        self.MERGE_SIZE = hf_config.vision_config.spatial_merge_size
+        self.IMAGE_FACTOR = self.PATCH_SIZE * self.MERGE_SIZE
```
[the remark this answers is not in the document]
Comment (mickqian) [reply python/sglang/srt/multimodal/processors/qwen_vl.py]: Could we apply this change?

File: python/sglang/srt/multimodal/processors/qwen_vl.py [hunk 2]
```diff
@@ -267,9 +239,19 @@ def __init__(self, hf_config, server_args, _processor, *args, **kwargs):
...
+
+        processor = self._processor
+        if hasattr(processor, "image_processor") and isinstance(
+            processor.image_processor, BaseImageProcessorFast
+        ):
+            self.MIN_PIXELS = processor.image_processor.size["shortest_edge"]
+            self.MAX_PIXELS = processor.image_processor.size["longest_edge"]
+        else:
+            self.MIN_PIXELS = 4 * 28 * 28
+            self.MAX_PIXELS = 16384 * 28 * 28
```
Comment (ZhengWG) [inline python/sglang/srt/multimodal/processors/qwen_vl.py]: video params may need to be reloaded too, such as:

```
video_size = getattr(video_processor, "size", {})
self.VIDEO_MIN_PIXELS = video_size.get("shortest_edge", 4 * 32 * 32)
self.VIDEO_MAX_PIXELS = video_size.get("longest_edge", 32 * 32 * 24576)
self.FPS = getattr(video_processor, "fps", 2)
self.FPS_MIN_FRAMES = getattr(video_processor, "min_frames", 4)
self.FPS_MAX_FRAMES = getattr(video_processor, "max_frames", 768)

```

Comment (yangsijia-celina) [reply python/sglang/srt/multimodal/processors/qwen_vl.py]: yes, I'm working on video part and I will create another pr to fix this part after thorough testing.

File: python/sglang/srt/multimodal/processors/qwen_vl.py [hunk 3]
```diff
@@ -267,9 +239,19 @@ def __init__(self, hf_config, server_args, _processor, *args, **kwargs):
...
-        self.MAX_PIXELS = 16384 * 28 * 28
+        self.PATCH_SIZE = hf_config.vision_config.patch_size
+        self.MERGE_SIZE = hf_config.vision_config.spatial_merge_size
+        self.IMAGE_FACTOR = self.PATCH_SIZE * self.MERGE_SIZE
+
+        processor = self._processor
+        if hasattr(processor, "image_processor") and isinstance(
+            processor.image_processor, BaseImageProcessorFast
+        ):
+            self.MIN_PIXELS = processor.image_processor.size["shortest_edge"]
```
Comment (yhyang201) [inline python/sglang/srt/multimodal/processors/qwen_vl.py]: Should we delete these variables that won’t be used?

Comment (yangsijia-celina) [reply python/sglang/srt/multimodal/processors/qwen_vl.py]: done
